Horus widget batt check updates

offer-shmuely opened the pull request:
1. calculate all values only once per 'refresh' to reduce CPU utilization
2. on battery disconnect from plane: 
        keep the data presented on screen (min/max/average)   
        non-active data is marked as blink

offer-shmuely commented:
![image](https://user-images.githubusercontent.com/7026911/57200341-6fea1c00-6f93-11e9-931f-412c5b529e6e.png)

bsongis committed:
Gyro added, dead code removed

bsongis committed:
CZ translations continued

bsongis committed:
Missing translation

bsongis committed:
Translations again

bsongis committed:
Assert added

bsongis committed:
Segfault removed

bsongis committed:
Compilation fix

bsongis committed:
Cosmetics

bsongis committed:
More work for X-Lite S/PRO

bsongis committed:
GyrX / GyrY eeprom support on X-Lite S/PRO

bsongis committed:
[X3] Only 5 switches

bsongis committed:
Renaming PXX / ACCESS modules constants

bsongis committed:
ACCESS started in Companion

kilrah committed:
Remove obsolete option

bsongis committed:
As X3 is X9LITE PCBX9 should be defined

3djc committed:
rename x3 to x9lite

3djc committed:
rename x3 to x9lite

bsongis committed:
Compilation fix

bsongis committed:
Compilation fix

bsongis committed:
Logical issue: X9DP is NOT an X9D. Will be fixed better when we introduce radio family definitions

bsongis committed:
Logical issue: X9DP is NOT an X9D. Will be fixed better when we introduce radio family definitions

3djc committed:
cosmetics

3djc committed:
cosmetics

bsongis committed:
New bunch of translations

bsongis committed:
[X3] missing MIXSRC line (+ an assert to avoid it happens again)

offer-shmuely commented:
lshems commented that it is bad practice to use the global parameters,
it interrupt when using more than one instance of the widget (and I use 3-4 on any of my models)
working on updating the code....

bsongis committed:
Renaming continued

bsongis committed:
Compilation fix

bsongis committed:
[X7] Bootloader fix

bsongis committed:
[X9 Lite] Simulator keys fixed

bsongis committed:
[X7] Simulator keys fixed

bsongis committed:
Update CREDITS.txt

kilrah committed:
VC++ compilation

bsongis committed:
3djc/x7 additional switches (#6422)

X7, add two additional switches

bsongis committed:
Remaining X3 in code

bsongis committed:
Bug with [EXIT Long] in TX / RX Options screen

bsongis committed:
Compilation fix

bsongis committed:
Cosmetics

bsongis committed:
Cosmetics

bsongis committed:
Cosmetics

bsongis committed:
[X10] Freeze issue fixed

bsongis committed:
Cosmetics

3djc committed:
Allow CROSSFIRE serial telemetry mirror

bsongis committed:
HORUS fixes

bsongis committed:
CZ next round of translations

bsongis committed:
HORUS fixes

3djc committed:
Horusgets 60 sensors, not 64

3djc committed:
Build simu for new radios

bsongis committed:
Compilation options updated

bsongis committed:
Compilation options updated

bsongis committed:
Compilation options updated

bsongis committed:
Cosmetics

3djc committed:
Verbose nighlies

kilrah committed:
Compilation fix

3djc committed:
Fix macosx compile

3djc committed:
Fix macosx compile

3djc committed:
Fix macosx compile

bsongis committed:
Compilation options updated

bsongis committed:
Windows compilation fight

3djc committed:
Allow accst protocols on xlites

bsongis committed:
Windows compilation fight

bsongis committed:
Revert "Allow accst protocols on xlites"

This reverts commit 5040e518

bsongis committed:
Revert "Compilation options updated"

This reverts commit 68a9edc8

bsongis committed:
Windows compilation fight

bsongis committed:
Windows compilation fight

bsongis committed:
Should work if we compile with -DWIN_DO_FULL_INSTALL=YES -DSIMU_TARGET=NO

bsongis committed:
License.txt to ISO-8859-1

bsongis committed:
Windows compilation fight

bsongis committed:
Windows compilation fight

bsongis committed:
Windows compilation fight clean

bsongis committed:
Windows compilation fight next round

bsongis committed:
Windows compilation fight next round

bsongis committed:
Windows compilation fight next round

bsongis committed:
Windows compilation fight next round clean

bsongis committed:
External module list depends on compilation options
- option PPM=NO is added for radios which don't support PPM anymore
- option SBUS=NO
- option DSM2=NO

bsongis committed:
Fixes in channels order in the Multi wizard

bsongis committed:
Options for building FrSky release firmware updated

bsongis committed:
Translations fix

bsongis committed:
Translations fix

bsongis committed:
Quel talent !

3djc committed:
Update firmware stamps asap

3djc committed:
small companion fixes

3djc committed:
Fix #•6425

3djc committed:
cosmetics

3djc committed:
Fix eeprom load multiples warnings

bsongis committed:
Untranslated strings splitted from others

3djc committed:
cosmetics

bsongis committed:
ACCST protocol support added

bsongis committed:
Strings big cleaning session (#6426)

Strings cleaning session

bsongis committed:
Companion new error detected on flash (when no device found)

bsongis committed:
Compilation fix

bsongis committed:
X10 extra switches sticks (#6423)

[X10] 2 extra sticks switches + 2 extra pots

bsongis committed:
String change

bsongis committed:
String change

bsongis committed:
String change

bsongis committed:
String change

bsongis committed:
Compilation fix

bsongis committed:
FourCC missing for new boards in Companion

bsongis committed:
Will raise a warning next time we forget a FourCC or an EEPROM size

bsongis committed:
Compilation fix

bsongis committed:
Compilation fix

bsongis committed:
Compilation fix

bsongis committed:
Compilation fix again (no possible to avoid it with VC++ ???)

bsongis committed:
VC++ warnings fixes

Commit:
Update gui_common.cpp

Added:
 - BAYANG DHD_D4
 - MJXQ PHOENIX and Freq tuning for E010
 - E01X E016H
 - GD00X V1 and V2
 - BUGSMINI BUGS3H

bsongis committed:
ACCST on ACCESS (#6430)

ACCST on ACCESS

ku2works committed:
updated japanese translation. (#6429)

3djc committed:
Fix ALT displayed in TopBar in FAI mode (2.3 version) (#6367)

SHezinger committed:
Some german translations added or corrected (#6344)

Some german translations added or corrected

Commit:
Wrong search & replace

bsongis committed:
Fix ported from #6331

bsongis committed:
#6331

3djc committed:
Prevent dist calculation overflow

3djc committed:
Improve nightly outputs

3djc committed:
Fix GPS overflow

3djc committed:
Make nightlies process more robust

3djc committed:
Make nightlies process more robust

kilrah committed:
[companion] Fix Bluetooth fields availability, add ownerId and gyro

3djc committed:
Make nightlies process more robust

kilrah committed:
Reserve gyro data for Horus

kilrah committed:
Compilation fix

3djc committed:
Multi compile fix

3djc committed:
日本語オプションを追加

3djc committed:
Add japaneese translation file

3djc committed:
Temporary fix for Sky9X compilation issue

3djc committed:
Use the Force, Luke

bsongis committed:
SKY9X-REVA removed as it doesn't even build since 2.2

bsongis committed:
Compilation fix

bsongis committed:
Unused file removed

bsongis committed:
Removing PPM support also removed Trainer / Trainee feature

3djc committed:
CROSSFIRE is no more a default, but firmware option (saves RAM)

3djc committed:
CROSSFIRE module availability is based on sport sharing

Commit:
Companion enhancements for checklist editor, model print and compare (2 3 branch) (#6434)

* Model print and compare include GFs enhancement

* Model print and compare include checklist contents

* Add missing brackets and code tidy up

* Print message if cannot open or read checklist file

* Missing GV label translation
Move some checklist code to helpers

* Refactor to use helpers and add cursor position

* Move improved version of removeAccents to helpers and apply to checklist

* Force fixed width font for edit window

bsongis committed:
Fixes #6439

bsongis committed:
[ACCESS] More bits for the RF protocol enum

bsongis committed:
Taranis X9D REV4a support removed (didn't compile since 2.2)

offer-shmuely commented:
This fix also issue:6418

bsongis committed:
Problem in Failsafe HOLD / NONE strings

3djc committed:
Allow PXX1 only

3djc committed:
Fix wrong BT password
Thix fixes #6440

3djc committed:
REvert previous fix and do it differently

Commit:
Fixes missing or invalid SD path (#6441)

bsongis committed:
XLite EEPROM 218 import fix

3djc committed:
Improve massively splash screen ! Thx MHotar !!!

3djc committed:
Fix fiel permissions

bsongis committed:
CMake options printed when building a firmware

bsongis committed:
Companion import fix

bsongis committed:
O9X FourCC support removed

3djc committed:
Fix companion import
This fixes #6443

bsongis committed:
Fix for negative gvars precision - inspired by #6446 thanks @ elecpower

Commit:
Add mouseover tool tips to label fields with context menus (#6448)

bsongis committed:
Aie

kilrah committed:
Fix X10 analogs

bsongis committed:
Merge bitfield macros

3djc committed:
Fix PPMUS

3djc committed:
Add Jumper t12 support by 3djc (#6451)

* Jumper T12 - companion part

* more work

* Add definitions

* fix libsimulator key handling

* Fix simu keys handling

* Cosmetics

* Continued

* Fix wrong regplace

* Fix wrong regplace

* Please travis

* Please travis

* Add T12 to Travis
Switch naming and display

* Cosmetics

* Work on modules

* Work on modules

* typo

* Fix navigation

* Cosmetics

* Introduce NAVIGATION_type

* Fix

* Make NAVIGATION_type more future proof

* Lots of cleanups and improvements

* Cosmetics

* Fix internal module menu

* Cosmetics

* Fix screen reversed

* Small fixes

* Fix navigation issue

* Please travis

* Fix LCD contrast

* Fix T12 switch layout

* Fix T12 keys screen and all radios keys translations

* Complete rework of SWITCHES screen for all radios

* comsetics

* VC++ compilation

3djc committed:
Prep for switch refactoring

3djc committed:
Typo

3djc committed:
Fix missing

3djc committed:
Switch refactoring continued

3djc committed:
Fix duplicate

bsongis committed:
Too long string on 128x64 screens

kilrah committed:
Properly disable BT when module not present

(cherry picked from commit 2bc9a00f32a29a085e4764042ffc7d61019944f1)

3djc committed:
X12S has Gimbal switchs handling as well

bsongis committed:
Bsongis/power control chip (#6445)

FRSK files support added

bsongis committed:
Now module / receiver / sensor firmware update from .frsk files should be OK

bsongis committed:
Compilation fix

3djc committed:
Fix X7 switches on main view

3djc committed:
Switches refactoring (#6454)

Switches refactoring

bsongis committed:
CZ translations

bsongis committed:
Bsongis/accst support over pxx2 (#6455)

ACCST support over PXX2

bsongis committed:
Refactoring

bsongis committed:
Cosmetics

bsongis committed:
Compilation fix

bsongis committed:
Compilation fix

bsongis committed:
Unused code removal

bsongis committed:
Compilation fix

bsongis committed:
X9 Lite powering issue

bsongis committed:
warning removed

bsongis committed:
R9M ACCESS / ACCST refactoring

bsongis committed:
Update build options

3djc committed:
Introduce AUX_SERIAL (#6456)

bsongis committed:
[XLiteS / X9 Lite] Will only propose "ACCESS / ACCST D16 / ACCST LR12" (no ISRM)

LapinFou committed:
Initial version

bsongis committed:
Module names constants / strings changed again

LapinFou committed:
Updated Companion French translation for OpenTX 2.3

bsongis committed:
R9M bugfix

bsongis committed:
R9M improvements

3djc committed:
Clean outdated and unused pxx2 code (#6458)

Remove outdated PXX2 code from X10 / X12S

bsongis committed:
Bsongis/menus refactoring (#6463)

Menus refactoring

bsongis committed:
Update CREDITS.txt

bsongis committed:
R9M warning (#6464)

R9M warning + refactoring

bsongis committed:
String change

MRC3742 committed:
Multi protocol updates to MPM version 1.2.1.51 (#6466)

MRC3742 committed:
Travis build error fix for PR #6466 Update MPM for 2.3

3djc committed:
Remove non working Theme

bsongis committed:
Update CREDITS.txt

bsongis committed:
Negative test inversion. Internal module mode displayed wrongly on X9 Lite

bsongis committed:
ACCST D16 / ACCESS mode removed until it is ready

bsongis committed:
Same shutdown animation on all boards

bsongis committed:
ACCST D16 is OK on ACCESS modules now

bsongis committed:
Fixes #6459

bsongis committed:
Switches position was not optimal in Main view [Switches]

kilrah committed:
Xlite has no trainer port

Commit:
Fixes 5602 Special Functions Inc/Dec Gvar (#6442)

Fixes #5602 (2.3)

bsongis committed:
Cosmetics

shellixyz committed:
Add vario sensor to CRSF telemetry

3djc committed:
T12 to use 9x navigation + navigation refactoring  (#6474)

* Move T12 to more suited 9x navigation

* Navigation refactoring

* Navigation refactoring continued

* Channels count in ACCESS mode fix (24 channels max)

bsongis committed:
Cosmetics

bsongis committed:
Radio / Version: FCC / EU variant displayed

bsongis committed:
BIND in ACCST mode didn't take care of 8CH / 16CH

bsongis committed:
Cosmetics

3djc committed:
Remove 1ms for mixer to pulse XJT (now 2ms)

bsongis committed:
Refactoring around trainer

bsongis committed:
ACCESS module type added to Companion

bsongis committed:
Compilation fix

bsongis committed:
Compilation fix

bsongis committed:
Cosmetics

3djc committed:
Fix critical issue when flashing 2.2.x firm from 2.3 bootloader (making 2.3 BL init wrong interupt not handled by 2.2 firm)

bsongis committed:
Cosmetics

bsongis committed:
Rename serial2 to aux_serial

bsongis committed:
Compilation fix

bsongis committed:
Compilation fix

bsongis committed:
ACCESS support started in Companion

bsongis committed:
Renaming "Power control chip" to "Power management unit"

bsongis committed:
Fix switch warning for X9 Lite

bsongis committed:
Always use 230400bauds for PXX2 on external module

3djc committed:
Improve switch layout display

3djc committed:
Companion T12 support (#6480)

* Introduce T12 own variant, radio side

* T12 Companion support continued

* Remove warnings

* Cosmetics

3djc committed:
Please Travis

bsongis committed:
Cosmetics

bsongis committed:
Fix an issue in statistics display

bsongis committed:
Cosmetics

bsongis committed:
Update CREDITS.txt

bsongis committed:
R9M subtype selection fix

bsongis committed:
Switches fix (because we have more than 256 switches today)

bsongis committed:
CZ translations

bsongis committed:
X7 didn't have the S.PORT connector enabled

3djc committed:
Fix nightlies compilation and add multi to Travis (it is mainstream after all)

bsongis committed:
S.PORT Flash fix

bsongis committed:
Lua warning popups could not be canceled!

bsongis committed:
Bsongis/radio tools (#6484)

Lua tools now displayed in RADIO / TOOLS
Add a function to write CENTERED text with lcdDrawText

bsongis committed:
PXX synchro using Hearbeat (#6482)

* [PXX] Latency using heartbeat sync optimized for X7 (stable @ 12ms)
* [PXX] Latency using heartbeat sync optimized for XLite Pro (stable @ 9ms)

Commit:
Add RX numbers to CPN models list (2.3) (#6447)

Add RX numbers to models list (#5088)

bsongis committed:
Typo

offer-shmuely commented:
Updated layout of the "1/4 screen size"
Now it use better the real estate, so additional data can be displayed even at this 1/4 size
:-)

I find it useful, since mostly I use the 1/4 screen with 2 BattCheck (one for each battery) + timer + LowestCellV

Hope other will find this usefull as well

![image](https://user-images.githubusercontent.com/7026911/62002538-f9326a00-b10e-11e9-8aaa-74d7bc602bdf.png)

Commit:
Merge remote-tracking branch 'origin/2.3' into 2.3

Commit:
Merge branch '2.3' into horus_widget_batt_check_updates

Commit:
cleaner look (optimize additional info)

offer-shmuely commented:
additional changes fro cleaner look

![image](https://user-images.githubusercontent.com/7026911/65992120-475d4400-e497-11e9-96d1-8c6ea08609bc.png)

Eldenroot commented on radio/src/translations/cz.h.txt:
```diff
@@ -833,15 +833,15 @@
+#define TR_NORMAL                      "Normální"
+#define TR_NOT_INVERTED                "Neinvert"
+#define TR_NOT_CONNECTED               "Nepřipojen"
+#define TR_CONNECTED                   "P5ipojen"
```
it should be "Připojen"

offer-shmuely replied:
looks that your changes are not relevant to this pull-request
please check them

offer-shmuely reviewed (commented):
.

offer-shmuely commented:
waiting for comments (or a merge)

3djc commented:
I get an error when trying to use it on top bar
![image](https://user-images.githubusercontent.com/5167938/69939619-a0ba1180-14e0-11ea-827b-81d296a1ef6c.png)

Commit:
fix index-out-of-bound when trying to use it on top bar

offer-shmuely commented:
thanks!!
fixed the error.
more comments please...

Commit:
cleaner code

Commit:
Keep historical info of lowest cells

3djc commented:
![image](https://user-images.githubusercontent.com/5167938/70027289-189c4080-15a2-11ea-83ff-03e5a6b432ec.png)

In top bar, vertical allignement doesnt seem to be ok

Commit:
BattCheck, update view in top-bar

offer-shmuely commented:
view in top-bar updated
thank you for your comments!!
Can I get more of them please?

![image](https://user-images.githubusercontent.com/7026911/70074080-ba537a00-1602-11ea-8571-ac7e09318929.png)

dmyers7 commented:
Thanks for working this.  This has driven me nuts for years.  The widget would more often than not crash when a full battery is plugged into the FrSky Batt Cell sensor and I could never figure out why and would fly with just the overall voltage on another widget.  Just grabbed the script from the Git as I wait for the SD Card contents to be updated on OpenTX.  Will give it a go in a bit with a fully charged battery.  Scripted compiled and started fine.  Topping off a battery now.  BTW, the RC1 notes for 2.3.3 mentioned that this has been released at part of the new SD Card package but all the packages are still dated from Nov 16 and the BattCheck Main.lua does not reflect the changes. Interestingly, even with a slight over charge to insure a little over 4.2V / cell the FrSky Cell Sensor (SP-FLVS) clamps the reading at 4.2V until the voltage drops below 4.2V.  Anyhow, all good and not crashing now; thanks again.

3djc commented:
It should be included in next RC. As you can see, it is still opened

offer-shmuely commented:
I am happy that it’s help
I will be happy to get more feedback and suggestions for new additional updates

For example, I am not happy with meaning of the color of each cell

3djc commented:
We have merged it for inclusion in 2.3.3RC2

salarian commented:
Is it possible to make this widget work with the VFAS (i.e. the battery voltage from FrSky telemetry)?

3djc commented:
Vfas does not contain individual cell data, which this plugin is all about

salarian commented:
Ah, I see. It makes sense. Thanks for quick response. The thing is this plugin is included in 2.3.3 for Jumper T16 as a standard widget. So as an end-user, once I saw it there I assumed it would work with any telemetry system...
Probably the description of the widget should be updated to mention it would only work with FLVS.

offer-shmuely commented:
I believe that the sensor supply single voltage? (i don’t have one)
Can you what you expect to see? 
Maybe it can be adapted

salarian commented:
Given that **VFAS** is a single number, probably the desired outcome would be to show the battery's charge percentage. As far as cell voltage goes, since individual cell voltages are not available, only an average should be shown. There are two ways to get the average cell voltage: 
- to use another telemetry variable (for example, FrSky telemetry typically has **VFAS** for total and **A4** for cell voltage).
- to estimate it from the total voltage. Betaflight e.g. uses a configurable divisor and simply divides the total battery voltage by it.

offer-shmuely commented:
Hi salarian
I prepare a test script that support single-voltage telemetry
do you  like to test it? (I do not have any VFAS)
how can I send you?

3djc commented:
Just as a comment, we should not compute cells value based on total voltage and estimated cell count. Either we have an accurate data and we can use it, or we dont. There is no issue to make a simpler vfas widget that translate total voltage to %,  without the cells portion

offer-shmuely commented:
Why?
we can work on "perfect" mode for the FVLSS
and for VFAS/A1/A2 we can work on simpler mode that asume a lipo, i.e. per cell voltage is 3.0 to 4.2
and calculate the percent.

I can create a different Widget for the A1/A2/VFAS, but I afraid it will harder for the user to choose the correct widget

What do you say?

3djc commented:
Because it is plain wrong to lead the user to belive all its cells are perfectly match and none is low. We either have accurate data, and we can do things with it, or we don't, and it's better not to display it. It can be same widget and you just skip the whole individual cell display

offer-shmuely commented:
ok, I will remove the individual cells display.

asalarian commented:
> 
> 
> Hi salarian
> I prepare a test script that support single-voltage telemetry
> do you like to test it? (I do not have any VFAS)
> how can I send you?

I am on vacation and traveling this week. I'll be back by the weekend though. I'll be happy to try the patch then.

offer-shmuely commented:
https://github.com/offer-shmuely/opentx/tree/shmuely/horus_batt_widget_support_single_cell/radio/sdcard/horus/WIDGETS/BattCheck

asalarian commented:
> https://github.com/offer-shmuely/opentx/tree/shmuely/horus_batt_widget_support_single_cell/radio/sdcard/horus/WIDGETS/BattCheck

Hello @offer-shmuely 

Sorry that I "disappeared" for a long time!
I tested the script with a few quads and few battery packs and it works great! 

The only thing I can see is a bit of a text that is appearing just besides the widget. Is it intentional (e.g. for your debug purposes?)

![screen](https://user-images.githubusercontent.com/10533408/79135086-15fef280-7daf-11ea-882d-2b9bcc38e185.png)

Thanks a lot!

offer-shmuely commented:
yes, it's a debugging print of the reset-event-detections
I fixed it.
however:
Since I accepted @3djc suggestion, we are NOT going to embedded this functionality into the official BattCheck on the official versions

So I changed the name of the widget to BattVFAS
Please use the official one for multi-cell sensors
and the BattVFAS only for single cell sensors

Enjoy


https://github.com/offer-shmuely/opentx/blob/shmuely/horus_batt_widget_support_single_cell/radio/sdcard/horus/WIDGETS/BattVFAS/main.lua
